components/data-provider/src/data_provider.rs

```diff
@@ -0,0 +1,162 @@
+use crate::cloneable_any::CloneableAny;
+use crate::data_entry::DataEntry;
+use crate::data_key::DataKey;
+use icu_locale::LanguageIdentifier;
+use std::any::Any;
+use std::any::TypeId;
+use std::borrow::Borrow;
+use std::borrow::BorrowMut;
+use std::borrow::Cow;
+use std::fmt;
+
+// Re-export Error so it can be referenced by "data_provider::Error"
+pub use crate::error::Error;
+
+/// A struct to request a certain hunk of data from a data provider.
+#[derive(PartialEq, Clone, Debug)]
+pub struct DataRequest {
+    pub data_key: DataKey,
+    pub data_entry: DataEntry,
+}
+
+impl fmt::Display for DataRequest {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}/{}", self.data_key, self.data_entry)
+    }
+}
+
+/// A response object containing a data hunk ("payload").
+#[derive(Debug, Clone)]
+pub struct DataResponse<'d> {
+    pub data_langid: LanguageIdentifier,
+    payload: Cow<'d, dyn CloneableAny>,
+    // source: Cow<'static, str>,
+}
+
+// TODO: Should this be an implemention of std::borrow::Borrow?
```

Comment:
I think you'd want `AsRef`. The difference between `Defer`, `Borrow` and `AsRef` is nuanced, but I see them as clearly distinct. In this case, `Borrow` is more for *accepting* either owned or borrowed arguments, `AsRef` is for exposing something as a reference to something (so, in this case, exposing a reference to a payload).

Reply:
This TODO is obsolete.  These functions are sufficiently complex that they should be their own functions.  I removed the TODO.